### ISSUE TITLE
[feature] Add a "global" or "once" flag to array and text define objects

### ISSUE THREAD

**Issue by jamshark70** [pure-data/pure-data#1370]
Here is something I have struggled with in a couple of contexts.

Let's say I want to create an abstraction, which requires global storage to be shared across instances. That could be an array, or let's say for this example that it's a [text] object being used like a hash table: send a symbol into an abstraction inlet, and it will [text search] etc.

Remember that the intention here is for a global hash table. If the abstraction includes [text define $0-table], then each instance has a distinct table. So using the instance number to avoid name collisions is not valid for this use case.

If the abstraction contains [text define myHashTable], then we will get "multiply defined" errors for every attempt to manipulate the table.

It would be error prone to ask the user to define the text object (and shouldn't it be the abstraction's job to manage its own resources?).

A clever forum user came up with a workaround: distinct [text] objects per instance, but [send]/[receive] pairs with shared names for manipulation. Weeelllll... oh-kayyyyy... but that's pretty ugly, and I'm gonna need another scotch to understand it a year down the road.

Wouldn't it be nice if one could write, for instance, `[text define -once myHashTable]`? For the first abstraction instance to be created, it would see that there is no `myHashTable` and make storage for it. For the second instance, it could simply point to the first: no "multiply defined" but the tables are shared.

**Comment by umlaeute** [comment]
i think a better approach (that overloading specific objects with a `-once` functionality) would be to have a defined way to create a *singleton* abstraction.

it's already possible to do that by abusing undocumented and private dynamic patching.


apart from that: what if you want to have two (or more) *independent* instances of the global hash-table?
the typical approach here is to use `[text define $1-myHashTable]` and pass `$0` (or whatever) to each abstraction instance that want's to use he global variable.

this way you can build a "namespace"-like thingy.

**Comment by Lucarda** [comment]
@jamshark70 

I'm not sure if I understand.

Why you **define** _array_ or _text_ in each abstraction. You have to define them only once and outside the abstractions. Then use _set_ or _get_ in all other abstractions.

**Comment by jamshark70** [comment]
> it's already possible to do that by abusing undocumented and private dynamic patching.

I.e. out of the reach of mere mortals. Make the functionality accessible.

> Why you define array or text in each abstraction. You have to define them only once and outside the abstractions. Then use set or get in all other abstractions.

Because the [text] or [array] is an implementation detail of the abstraction. If I publish my abstraction, then every user of that abstraction has to be aware of internal implementation details, which is ridiculous. Abstractions should hide those details.

**Comment by Lucarda** [comment]
How would Pd hide anything?

You can place the **define**s in another abstraction?

**Comment by jamshark70** [comment]
@umlaeute 

> that's why i was suggesting to provide a mechanism to create singleton abstractions. this seems to be the exact solution you are looking for (in a much more generalized and useful way), no idea why you don't acknowledge this.

The reason is because I was about 25% paying attention to this conversation and 75% on other things... which probably means I should have withheld the last reply until I could do it absolutely completely. But I do think we should sometimes not assume that everybody's reply is necessarily complete.

Anyway... yes. I would be OK with a way to have a singleton abstraction.

@Lucarda 

> How would Pd hide anything?

I'm talking about (and umlaeute is too) one of the core principles of object-oriented design: the distinction between public interface and private implementation.

If one is using the abstraction mechanism to design a new object (which is one of the intended uses), then:

- The abstraction may do *anything* that it needs to do internally, to get the result.
- The user *should not be aware of those internals*. This is encapsulation. The object owns its own stuff. You shouldn't have to tinker with the object's internal stuff just to get a basic result. If the object requires a storage table, and if the user is responsible for creating that storage table, then encapsulation is broken. This becomes just one more thing that the user might mess up = the object is unreliable.
- The object should define a public set of messages which it will accept (in Pd, through inlets) and a public set of expected responses to those messages. The user should need to think 100%, *only*, in terms of the public messages going in and public replies coming out. This is the public interface.

If Pd makes it prohibitively difficult to implement, for instance, a global lookup table shared across multiple instances of an abstraction, then Pd's object-oriented design would have to be considered weaker than other programming environments' OO design. In SC I can write a class `MyLookupThingy { classvar hiddenGlobalTable; *initClass { hiddenGlobalTable = IdentityDictionary.new } .... more stuff }` and it's super easy.

Singleton abstractions would go a long way towards plugging this particular hole.

**Comment by Spacechild1** [comment]
> for instance, a global lookup table shared across multiple instances of an abstraction

I was just about to bring this up. Lookup tables are the prime example for a private resource that is shared between instances and that the user shouldn't have to know about.

Anyway, I think we all seem to agree that we need such a mechanism and that singleton abstractions would be a good general solution.

**Comment by umlaeute** [comment]
in general i agree with @jamshark70 

however, i would like to comment on this:
>     * If the object requires a storage table, and if the user is responsible for creating that storage table, then encapsulation is broken

in a way yes.
however, it's not uncommon to require a user to properly initialize the system.
E.g. if you want to use GStreamer in your (C-)application, you really [must add something like this](https://gstreamer.freedesktop.org/documentation/application-development/basics/init.html?gi-language=c) to the beginning of your code:

~~~C
gst_init (&argc, &argv);
~~~

similarily, you might require your users to create a single object `[fudel_init]` if they want to use your *fudel* library.
This `[fudel_init]` is then free to initialize whatever global hashtables it likes - and hide this fact from the user (encapsulation).

**Comment by Spacechild1** [comment]
> similarily, you might require your users to create a single object [fudel_init] if they want to use your fudel library.

I have used this approach myself. A potential problem is ordering. The user has to make sure that `[fudel_init]` is created before any other objects from the `fudel` library.

Regarding singleton abstractions: I have been wondering if we could maybe extend `[declare]` for this purpose. For example, `[declare -init fudel]` would look for a patch `fudel.pd` and create it (invisibly) if it hasn't been created already. Alternatively, we could override the `-lib` flag to hide the distinction between abstraction libraries and compiled libraries, so `[declare -lib fudel]` would either load a compiled library or a library patch.

**Comment by reduzent** [comment]
I agree that a singleton pattern would be generally useful, but also specifically for this case. One advantage is that the singleton survives deletion of nodes. With `[text define -once fudel]` you lose all data when deleting the instance holding the data (and there is no way to know which instance it is, unless some notification is implemented, too). 

Not that I'm against Pd providing a generic way to allow this, but actually it can be done already now with patching as @umlaeute pointed out. netpd has been doing this for years. Besides other things, it is used to create lookup tables that require significant amount of memory and that would be wasteful to create in each instance.

If Pd would provide an interface for managing singletons, I'd wish it would keep track of nodes using the singleton so that it could be deleted when no node is using it anymore. The tracking is something I believe is not so easy to implement in patching, at least not without polling. But then again, externals cannot also not be unloaded and we've been living with that just fine, so I'm not quite sure about this.

**Comment by umlaeute** [comment]
> If Pd would provide an interface for managing singletons, I'd wish it would keep track of nodes using the singleton so that it could be deleted when no node is using it anymore. 

right.
if we provide a means to do singletons, then it should be properly done.

> The tracking is something I believe is not so easy to implement in patching, at least not without polling. But then again, externals cannot also not be unloaded and we've been living with that just fine, so I'm not quite sure about this.

i don't think it would be much of a problem on the C side of things (it's much more complicated on a patch level).

anyhow: i think the request for a `[singleton]` object should go into a separate issue/feature request (with this one being too long already ;-))

**Comment by reduzent** [comment]
> anyhow: i think the request for a `[singleton]` object should go into a separate issue/feature request (with this one being too long already ;-))

Before adding noise to the next feature request, let me pollute this one with my thought.  [singleton] would have an interface similar to [clone]?

```
[singleton <abstractionname> <firstarg> <secondarg> ...]
```
For two different calls to be accounted for the same singleton instance, all arguments must be the exactly the same. If arguments differ, separate singleton instances are created. This would allow for the same singleton-abstraction to be used in different contexts simultaneously. Does that make sense?

**Comment by jamshark70** [comment]
@umlaeute
> i think the request for a [singleton] object should go into a separate issue/feature request (with this one being too long already ;-))

Sure, I can agree to close and make a new one... after:

@reduzent
> For two different calls to be accounted for the same singleton instance, all arguments must be the exactly the same. If arguments differ, separate singleton instances are created.

That's a legit question. If arguments are allowed at all for a singleton, there is always the prospect of conflicts.

One solution is to enforce it as a true singleton: one and only one instance, in which case arguments become pointless (so there would be no need to support them).

But umlaeute is interested in "what if you want to have two (or more) independent instances of the global hash-table?" ... which is technically no longer a singleton, even though a lot of resources just call it a singleton anyway.

From [Wikipedia on "multiton" objects](https://en.wikipedia.org/wiki/Multiton_pattern), it seems that the most common strategy to disambiguate multiton instances is to refer to them by name... so the choices are:

- Let the [singleton] object assign a key, and allow arguments. But that really just pushes the problem down a level.
  - If there's [singleton xyz myKey10 10] and [singleton xyz myKey100 100], that should be OK because the keys are different.
  - But: [singleton xyz myKey 10] and [singleton xyz myKey 100] would pose a problem.
- Automatically construct a key from arguments. This gets tricky. Simple concatenation could result in a very long key. Hashing gets messy (I saw this happen within a SuperCollider extension framework... if it's not done very carefully, ohhhh how it can be breakable).
  - Or, another way could be a multilevel hash table -- if it's say [singleton xyz 123 8 foo], then you would have 4 nested hash tables. Probably best.

Last -- while I understand the wish to generalize, and also to minimize demands on developer time -- keyed "singleton" abstractions really just push the keying requirement to a level away from the text or array objects. You couldn't simply create a shared text or array. You would have to create a separate file containing the e.g. [text define mything] and then, in the parent patch, a singleton referring to the separate file.

That is, it would *work*, but it's not convenient. Technical requirements impede the user's creative flow. This is the kind of thing where you think "oh, that's not too bad" but then, after you've had to do it a hundred times, you think "why the [expletive] could they not have just put it into the text/array where I really need it? Why do they force me to make this extra layer that has no other purpose at all?"

I mean, I know which direction the feature is going to go, but... usability really does matter.

**Comment by umlaeute** [comment]
> One solution is to enforce it as a true singleton: one and only one instance, in which case arguments become pointless (so there would be no need to support them).

well yes or no.
this boils down to *either* (XOR):
- `[singleton xyz A 2 3]` and `[singleton xyz B 20 30]` to refer to different objects
- mandate a something like `[singleton xyzA]` and `[singleton xyzB]` where the *xyzA.pd* would contain a single object `[xyz A 2 3]` and the *xyzB.pd* would contain a single object `[xyz B 20 30]`

this looks like overkill to me, for no real benefit (apart from having a notion of doing a "true object-oriented singleton"

> - Automatically construct a key from arguments. This gets tricky

i obviously haven't thought much about the implementation yet, but i don't really see a problem here.

Pd being an instance-oriented language with a clear distinction between setup time and run time, and patches really only getting a handle (at runtime) to the actual object (or a proxy thereof), i cannot think of a problem however long the key gets (and the hashing takes): hashing the key is only required to get the handle to the object and this is only happening in the life-time of a singleton instance.


> You couldn't simply create a shared text or array. You would have to create a separate file containing the e.g. [text define mything] and then, in the parent patch, a singleton referring to the separate file.

i'm not sure i understand this.

true, my `[singleton]` proposal follows the model of `[clone]` so you do need a wrapper abstraction, that might only contain `[text define mything]`.
this is a limitation that we also see with `[clone]` (and it might be interesting to allow `[singleton text define mything]` on it's own - but let's leave that for later...)

what i don't understand is your "parent patch" thing: i thought each of your consumers (e.g. abstractions, to be duplicated in your parent patch) of the global object would have their own `[singleton xyz]`

> "why the [expletive] could they not have just put it into the text/array where I really need it? Why do they force me to make this extra layer that has no other purpose at all?"

the key here is the continuation slash in "text/array": it does not end with those two objects (shared network connections,...).
the purpose of all this is that a user only has to learn a single way to create a global instance of anything, rather than having to wade through a handful of mind-boggling object-flags.
